Update Congestion Control section (#115)
quicwg/ack-frequency: draft-ietf-quic-ack-frequency.md

## Congestion Control

A sender needs to be responsive to notifications of congestion, such as
a packet loss or an ECN CE marking. To enable a sender to respond to potential
network congestion in a timely fashion, usually at least one acknowledgement
per round trip is needed if there are unacknowledged ack-eliciting packets
in flight. A sender can accomplish this by setting the Ack-Eliciting Threshold
to a value no larger than the current congestion window or the Request Max Ack

> Delay value to no more than the estimated round trip. A sender can typically

Reviewer: ```suggestion
Delay value to no more than the estimated round trip. Alternatively, a sender can
```

Reviewer: I also proposed this sentence based on the discussion. We don't really need it but, I think in this part of the draft a bit of a warning could still be good:

> However, the congestion window particularly but also the RTT are dynamic and therefore might require frequent updates if the selected value are close to this limit.

Author: That's true.  Do you think we need to explicitly say that?

accomplish this by sending an IMMEDIATE_ACK frame once per round trip, though
if the packet containing an IMMEDIATE_ACK is lost, detection of that loss will be
delayed by the reordering threshold or requested max ack delay.

> Note that it's possible the RTT is smaller than the receiver's timer granularity,

Reviewer: ```suggestion
Note that it is possible that the RTT is smaller than the receiver's timer granularity,
```

## Congestion Control

A sender needs to be responsive to notifications of congestion, such as
a packet loss or an ECN CE marking. To enable a sender to respond to potential
network congestion in a timely fashion, usually at least one acknowledgement
per round trip is needed if there are unacknowledged ack-eliciting packets
in flight. A sender can accomplish this by setting the Ack-Eliciting Threshold
to a value no larger than the current congestion window or the Request Max Ack
Delay value to no more than the estimated round trip. Note that the congestion

> window particularly but also the RTT are dynamic and therefore might require frequent
> updates if the selected value are close to these limits. Alternatively, a sender can
> accomplish this by sending an IMMEDIATE_ACK frame once per round trip, though
> if the packet containing an IMMEDIATE_ACK is lost, detection of that loss will be
> delayed by the reordering threshold or requested max ack delay.

Reviewer: ```suggestion
window particularly but also the RTT are dynamic and therefore might require
frequent updates if the selected value are close to these limits. Alternatively,
a sender can accomplish this by sending an IMMEDIATE_ACK frame once per
round trip, though if the packet containing an IMMEDIATE_ACK is lost,
detection of that loss will be delayed by the reordering threshold or requested
max ack delay.
```

## Congestion Control

A sender needs to be responsive to notifications of congestion, such as
a packet loss or an ECN CE marking.  To enable a sender to respond to potential

> network congestion, a sender SHOULD cause a receiver to send an acknowledgement

Reviewer 2: Why is it the sender's responsibility to ensure that the receiver sends acknowledgments?  I have a fair idea, but it needs a bit of explanation as it is not obvious.

My first thought was that maybe the receiver shouldn't accept a delay that exceeded the RTT.  That might still be a good safety measure, even if the view of RTT is subjective and it might disagree with the sender on how that limit applies.

Author: There are cases where a delay larger than RTT, at least min RTT could be appropriate.  One example is when RTT << timer granularity, ie: 10 us network latency and 1ms timers in a datacenter.  Now that I think about it, I will call that case out.

Reviewer 3: I would expect you might do things differently in a DC -  we could declare a minimum here...

Reviewer: Yes, I'm also not sure about this case. Generally I think you should try to at least send one ACK per RTT. If you timer granularity is lower than RTT, you could still use a packet threshold to achieve this.

If you want to diverge from this guidance in a DC is something we don't really control. But I think at least once per RTT is the right guidance for the Internet.

Reviewer 3: I'm not sure why you suggest /one ACK per RTT/ ? In the measurements I have seen, several (e.g. 4) per RTT are beneficial to be avoid the effects of path variation and loss on the return path. But, agree on the backup of using the packet threshold to achieve this, which seems also to work well if the RTT varies.

at least once per RTT if there are unacknowledged ack-eliciting packets in flight.
A sender can accomplish this by sending an IMMEDIATE_ACK frame once per

> round-trip time (RTT), or it can set the Ack-Eliciting Threshold and

Reviewer 2: Setting the Ack-Eliciting Threshold has no effect on whether the receiver will acknowledge.  It only works if that many packets are sent, which is something that I don't think you can guarantee when setting the value.

Reviewer 3: I agree there is a bound on how long the transport should defer reacting when it has sent data - this is a control loop. For an Internet path, to me this is measured in RTTs, or part of an RTT. (I don't know whether we can state the same constraints to apply in both a heterogenous Internet and in a high speed DC). I was pretty sure that previous detailed analysis of feedback systems showed requirements for stable feedback at multiple control signals per RTT, so I'd encourage care here.

Reviewer 3: he /or/ is still problematic - it's not one or the other, because the time is what actually matters for protocol robustness; and the count threshold is really helpful when there are path characteristic changes, because it prevents very large numbers of packets per ACK - or we simply wish to reduce the rate. Perhaps two sentences would be wiser.

Request Max Ack Delay values to be no more than a congestion window and an
estimated RTT, respectively.

A congestion controller that is congestion window limited relies upon receiving
acknowledgements to send additional data into the network.  An increase in
acknowledgement delay increases the delay in sending data, which can reduce the

> achieved bandwidth.  Congestion window growth can also depend upon receiving

Reviewer 2: bandwidth or throughput?

Author: Thanks, throughput.

> acknowledgements, such as in slow start (Section 7.3.1 of {{QUIC-RECOVERY}}),

Reviewer 2: ```suggestion
acknowledgements, such as in slow start ({{Section 7.3.1 of QUIC-RECOVERY}}),
```

so delaying acknowledgements can delay the increase in congestion window.

Window-based congestion controllers, such as the one defined in {{QUIC-RECOVERY}},
rely on receipt of acknowledgments to send additional data into the network,

> so performance is degraded if acknowledgments are delayed excessively.

Reviewer 2: This looks like a repetition of the more precise text in the previous paragraph.

Author: Agreed, it's the leftover text from what I started with.  Should I just delete it?

Reviewer 3: To me though for Internet paths, the key issue is not the "performance" reduction, the primary concern of the transport area was the increased congestion risk - especially to other flows that share a common bottleneck along the path.

Author: What increased congestion are you concerned about by excessively delayed acks?

Reviewer 3: What I mean is a sluggish response to congestion when the sender was neither cwnd nor flow credit-limited (i.e. sender can still send without receiving feedback) and the ACK frequency is low (>>IW packets per ACK), where the sender persists sending new data into the network after the congestion event, i.e. CE-mark or loss.

Author: This paragraph is talking about the congestion controller potentially being CWND limited and not being able to send.  It's not discussing anything related to congestion (ie: CE or missing packets).

Reviewer 3: As long as the ID has the CC issues as well as the performance ones, I'll be happy.

Reviewer: I think risks about not receiving timely feedback are covered in the initial paragraph of this PR. Let's merge this for now and see if anything is still needed in addition.

## Congestion Control

A sender needs to be responsive to notifications of congestion, such as

> a packet loss or an ECN CE marking.  To enable a sender to respond to potential
> network congestion, a sender SHOULD cause a receiver to send an acknowledgement
> at least once per RTT if there are unacknowledged ack-eliciting packets in flight.
> A sender can accomplish this by sending an IMMEDIATE_ACK frame once per
> round-trip time (RTT), or it can set the Ack-Eliciting Threshold and
> Request Max Ack Delay values to be no more than a congestion window and an
> estimated RTT, respectively.

Reviewer: ```suggestion
a packet loss or an ECN CE marking. Therefore, {{Section 13.2.2 of QUIC-TRANSPORT}} requires
the receiver to "send an ACK frame after receiving at least two ack-eliciting packets".
This document overwrites this requirement by providing the sender a way to signal the needed
ACK frequency to the receiver. To enable a sender to respond to potential
network congestion in a timely fashion, usually at least one acknowledgement
per RTT is needed if there are unacknowledged ack-eliciting packets in flight.
A sender can accomplish this by sending an IMMEDIATE_ACK frame once per
round-trip time (RTT). Further, a sender SHOULD  NOT set the Ack-Eliciting Threshold and
Request Max Ack Delay values to be more than a congestion window and an
estimated RTT, respectively.
```

Reviewer 3: /the sender a way to signal/the sender with  a way to signal/

Reviewer: Ian and I discussed today about the new "SHOULD NOT" in the proposed text. During discussion we reached the conclusion that it's better to not use normative language here. however, the more I think about it, I think it's actually important to make a recommendation about the boundary on how to set the values.

However, maybe this is just not the right place in the document? This paragraph discusses how a data sender can use this extension to ensure that there is at least one ack RTT. However, if we want to have normative language, maybe this should be added to the ACK_FREQUENCY Frame section?

Reviewer: ```suggestion
a packet loss or an ECN CE marking. Therefore, {{Section 13.2.2 of QUIC-TRANSPORT}} 
indicates "A receiver SHOULD send an ACK frame after receiving at least two ack-eliciting packets".
This document overwrites this requirement by providing the sender with a way to signal its preferred
ACK frequency to the receiver. To enable a sender to respond to potential
network congestion in a timely fashion, usually at least one acknowledgement
per round-trip time (RTT) is needed if there are unacknowledged ack-eliciting packets in flight.
A sender can accomplish this by sending an IMMEDIATE_ACK frame once per
RTT. Similarly, setting the Ack-Eliciting Threshold to a lower value than the current congestion window or
the Request Max Ack to less than the estimated RTT are two options to request at least one ACK frame per RTT.
However, the congestion window particularly but also the RTT are dynamic and therefore might require frequent updates if the selected value are close to this limit.
```
This removes normative language and adds even more explanation... Sorry for making the text longer but given this is a discussion section, maybe more words actually have to make it clearer.

Note that it's possible the RTT is smaller than the receiver's timer granularity,
as communicated via the 'min_ack_delay' transport parameter, preventing the
receiver from sending an acknowledgment every RTT in time.  In these cases,

> 'Ignore Order' can be harmful, because it delays fast loss detection.

Reviewer: ```suggestion
'Reordering Threshold' can be harmful, because it delays fast loss detection.
```

> 

Reviewer 3: Indeed, deferring loss detection  impacts performance - I think it more important not to delay the congestion response.

Reviewer: I opened issue #138

A congestion controller that is congestion window limited relies upon receiving
acknowledgements to send additional data into the network.  An increase in
acknowledgement delay increases the delay in sending data, which can reduce the

> achieved bandwidth.  Congestion window growth can also depend upon receiving

Author: ```suggestion
achieved throughput.  Congestion window growth can also depend upon receiving
```

acknowledgements, such as in slow start ({{Section 7.3.1 of QUIC-RECOVERY}}),

> so delaying acknowledgements can delay the increase in congestion window.

Reviewer: and create bursts in no pacing is used.

Reviewer 3: I am unsure it is useful to say: /such as in slow start/ ... (isn't it actually any case where the sender is cwnd-limited?)

Reviewer: Kind of but in slow start is actually a noticeable problem because in congestion avoidance you only increase by something small like 1 MSS per RTT and it doesn't matter that much when exactly you increase the congestion window within the RTT.

Reviewer 3: Sure - if thresh is small then it's noticeable for CA. So, can we write e.g., this can be particularly significant in slow start...

> acknowledgements, such as in slow start ({{Section 7.3.1 of QUIC-RECOVERY}}),
> so delaying acknowledgements can delay the increase in congestion window.

Reviewer: ```suggestion
acknowledgements. This can be particularly significant in slow start ({{Section 7.3.1 of QUIC-RECOVERY}}),
as delaying acknowledgements can delay the increase in congestion window and
creating packet bursts if pacing is not used.
```

Window-based congestion controllers, such as the one defined in {{QUIC-RECOVERY}},
rely on receipt of acknowledgments to send additional data into the network,

> so performance is degraded if acknowledgments are delayed excessively.

Reviewer 3: Looking again the two are slightly different  and need to be combined.

Both are describing the case when the sender is window-limited - (1) it increases path delay, and hence lowers throughput (2) it delays growth of cwnd, and where performance can be impacted. (and we avoid anyone thinking fewer ACKs always means lower throughput).

Reviewer: I think this is well covered in the previous paragraph.

> 
> Window-based congestion controllers, such as the one defined in {{QUIC-RECOVERY}},
> rely on receipt of acknowledgments to send additional data into the network,
> so performance is degraded if acknowledgments are delayed excessively.

Reviewer: ```suggestion
```
